### PR TITLE
Units for BitOps.TrailingZeroCount

### PR DESCRIPTION
**Pull request by grant-d**
Units for `BitOps` from related PRs: [LZCNT, LOG2](https://github.com/dotnet/coreclr/pull/22497), [TZCNT](https://github.com/dotnet/coreclr/pull/22333), [ROTL/R, POPCNT](https://github.com/dotnet/coreclr/pull/22584)

cc @tannergooding, @jkotas

### CODE REVIEW

File: src/Common/tests/Common.Tests.csproj [hunk 1]
```diff
@@ -74,6 +74,7 @@
     <Compile Include="Tests\System\CharArrayHelpersTests.cs" />
     <Compile Include="Tests\System\IO\PathInternal.Tests.cs" />
     <Compile Include="Tests\System\IO\StringParserTests.cs" />
+    <Compile Include="Tests\System\BitOpsTests.cs" />
```
Comment (stephentoub) [inline src/Common/tests/Common.Tests.csproj]: Seems like you also need to include https://github.com/dotnet/corefx/blob/master/src/Common/src/CoreLib/System/BitOps.cs ?

Comment (grant-d) [reply src/Common/tests/Common.Tests.csproj]: Perhaps this is not the right project to add these tests. I thought that methods in CoreCLR/CoreLib would be visible here after the sync-up?

Comment (stephentoub) [reply src/Common/tests/Common.Tests.csproj]: The file gets mirrored to the repo, but you'd still need to include it in this test project.

Comment (grant-d) [reply src/Common/tests/Common.Tests.csproj]: Thanks. Fixed

**Comment by grant-d** [comment]
@stephentoub it's failing due to this line in the `CoreCLR` sources - I am not familiar with what that line imports. Any advice on how to fix this, perhaps that line can be removed? Or I need a suitable `#if`. Or I can even introduce a dummy namespace in the test project.

`...BitOps.cs(9,7): error CS0246: The type or namespace name 'Internal' could not be found...`

`using Internal.Runtime.CompilerServices;`

File: src/Common/tests/Tests/System/BitOpsTests.cs [hunk 2]
```diff
@@ -0,0 +1,63 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using Test.Cryptography;
+
+using Xunit;
+
+namespace Internal.Runtime.CompilerServices
```
Comment (grant-d) [inline src/Common/tests/Tests/System/BitOpsTests.cs]: `BitOps` imports this namespace (I am not sure why).
Is this workaround too hacky?

**Comment by grant-d** [comment]
Tests are failing due to this code in `BitOps`. It's missing the explicit conversion to `IntPtr`, which was apparently fine on Windows but not Linux et al.
I will fix in the next push I make in `coreclr`.
```csharp
public static int TrailingZeroCount(uint value)
    ....
    return Unsafe.AddByteOffset(
        ref MemoryMarshal.GetReference(s_TrailingZeroCountDeBruijn),
        (IntPtr)(((value & -value) * 0x077CB531u) >> 27)); // <--- (IntPtr)
```

**Comment by stephentoub** [comment]
> Any advice on how to fix this, perhaps that line can be removed?

Your workaround of adding an empty namespace to the test file seems fine.  Thanks.

File: src/Common/src/CoreLib/System/BitOps.cs [hunk 3]
```diff
@@ -55,9 +55,9 @@ public static int TrailingZeroCount(uint value)
 
             // uint.MaxValue >> 27 is always in range [0 - 31] so we use Unsafe.AddByteOffset to avoid bounds check
             return Unsafe.AddByteOffset(
-                ref MemoryMarshal.GetReference(s_TrailingZeroCountDeBruijn), 
+                ref MemoryMarshal.GetReference(s_TrailingZeroCountDeBruijn),
                 // Using deBruijn sequence, k=2, n=5 (2^5=32) : 0b_0000_0111_0111_1100_1011_0101_0011_0001u
-                ((uint)((value & -value) * 0x077CB531u)) >> 27);
+                (IntPtr)(int)(((uint)((value & -value) * 0x077CB531u)) >> 27)); // shift over long also expensive on 32-bit
```
Comment (grant-d) [inline src/Common/src/CoreLib/System/BitOps.cs]: I will revert this change in a subsequent commit. Just checking that this same fix in `coreclr` fixes the problem

Comment (grant-d) [reply src/Common/src/CoreLib/System/BitOps.cs]: Fix works. Reverting.

**Comment by grant-d** [comment]
Jumped the gun. These won't pass until the sync from `coreclr` has happened

File: src/Common/tests/Common.Tests.csproj [hunk 4]
```diff
@@ -67,10 +67,14 @@
     <Compile Include="$(CommonPath)\System\Security\IdentityHelper.cs">
       <Link>Common\System\Security\IdentityHelper.cs</Link>
     </Compile>
+    <Compile Include="$(CommonPath)\CoreLib\System\BitOps.cs">
```
Comment (grant-d) [inline src/Common/tests/Common.Tests.csproj]: We can remove this once BitOps is exposed

File: src/Common/tests/Tests/System/BitOpsTests.cs [hunk 5]
```diff
@@ -0,0 +1,312 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using Xunit;
+
+namespace Internal.Runtime.CompilerServices
```
Comment (grant-d) [inline src/Common/tests/Tests/System/BitOpsTests.cs]: We can remove this once BitOps is exposed

**Comment by grant-d** [comment]
@dotnet-bot test NETFX x86 Release Build please
@dotnet-bot test OSX x64 Debug Build please
@dotnet-bot test Windows x64 Debug Build please
@dotnet-bot test Windows x86 Release Build please

**Comment by grant-d** [comment]
@dotnet-bot test NETFX x86 Release Build please

File: src/Common/tests/Tests/System/BitOpsTests.cs [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+{
+    // Dummy namespace needed for compilation of BitOps
+}
```
Comment (grant-d) [inline src/Common/tests/Tests/System/BitOpsTests.cs]: We can remove this once BitOps is exposed

**Comment by grant-d** [comment]
@dotnet-bot test this please

**Comment by grant-d** [comment]
@dotnet-bot test corefx-ci (Linux x64_Release) please

**Comment by grant-d** [comment]
Can we merge?

**Comment by grant-d** [comment]
Thanks for approving Jan, LGTM?

**Pull request merged**